### DISCUSSION TITLE
Live image updates

### DISCUSSION THREAD

**Discussion by dg-pb** [zauberzeug/nicegui#1630] [Q&A]
### Question

Hello,

Thank you for a beautiful project. It looks nice. :)

Also, a question. I am planning to start working on UI for a project where RGB image needs to be updated in real time. Reloading the whole image is a bit too expensive, so I was wondering if there was some sort of communication channel between python and image widget to update only selected pixels.

Thanks again,
DG

**Comment by falkoschindler** [comment]
Thanks, @dgrigonis! 😊

There's currently no built-in support for updating images on a pixel-basis. In principle it could be done by implementing a custom UI element like we did for `ui.log` or `ui.scene` which update their content incrementally. But for images we would also need to find a suitable data format and client-side library...

Maybe it's easier to work with image tiles? That allows you to use standard elements like `ui.grid` and `ui.image` while still achieving quick updates.

**Reply by dg-pb** [reply]
It is a fairly rapid updating. Some images are 5000x5000 and update rate can be up to 20-30sec. It doesn't have to be updated every iteration, but at least every 10th. Also, converting them to smaller ones is also expensive. They are stored in numpy arrays and it is not as fast as image oriented C code.

Currently I am using bokeh to sketch things in jupyter. It does communicate with js side and can patch data without redrawing it. I am currently redrawing it every 10 iterations and it is close to being a bottleneck (at least in jupyter). Maybe it will work smoother with on a server. It works with RGBA 32bit integers btw.

Would it be possible to easily integrate bokeh into nicegui?

**Reply by falkoschindler** [reply]
See https://github.com/zauberzeug/nicegui/discussions/359 for a bokeh feature request.

I think I don't quite get the requirement: Updating a 5000x5000 image by converting it to a PNG, sending it to the client and rending it in a browser should be easily doable in 20-30sec. Or do you mean 20-30Hz? Oh and do you need to send it over the internet or local network?

**Reply by dg-pb** [reply]
Eventually I aim for it to be hosted. Work is done on a server side, so updates will have to travel to client over the internet. I mean 20-30hz. Essentially, the app I am making is somewhere in between image editor and image processing. It needs some basic drawing functionality with callbacks from client to server and back, but it doesn't need to be 0-latency updates as drawing actions are determined algorithmically on a server side and done in chunks (not manual drawing). So what I want is to give user visibility to the progress that is being done.

Thanks for link to bokeh thread.

**Reply by rodja** [reply]
5000x5000 is larger than the resolution of most displays. Pumping these in high frequency over possible bad internet connections is super tricky. I would think hard if a scaled-down version of the data may be generated by any means. I guess it's easier to write some c++ code and wrap it in Python than trying to make users internet connections fast enough.
Secondly you'll probably need to progressiv update only the parts of the image which change. Similar to what modern video codes do. Maybe there are libraries out there to archive this? Game platforms like Google Stadia or NVIDIA GeForce NOW are doing similar stuff.

**Reply by dg-pb** [reply]
I will experiment with a higher level tools for now. Bokeh seems to be working reasonably for updating 720x720 every at 1Hz. Maybe implementing efficient downscaling and "patching" data in json will be enough.

Here is an example of my requirement:
https://programforyou.ru/projects/string-art-generator

**Reply by rodja** [reply]
720x720 at 1Hz should also be working with `ui.interactive_iamge`. Actually, you can update faster. The client will automatically decide when to pull a new image. That way slow connections will have slow update while fast connections will have faster update rate.

**Reply by dg-pb** [reply]
Thank you, I will experiment with it for sure before getting to more complex solutions. Either this or bokeh first. If that is insufficient will see what else can be done.
